enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,213 @@
+	// When this list is empty, no identities are provisioned for users.
+	// +optional
+	IdentityProviders []IdentityProvider `json:"identityProviders,omitempty"`
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+
+	Audit OAuthAudit `json:"audit"`
+}
+
+type OAuthAudit struct {
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+  // This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+### As an administrator I want to inspect successful login and login failure attempts
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs.
+
+### As an administrator I want to inspect logout events and the deleted user
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect apiserver audit logs for deleted `oauthaccesstoken` resources.
+Audit log entries include the deleted user.
+
+### As an administrator I want to inspect the source IPs
+
+## Design Details
+
+### Successful Login and Login failure events
+
+To create successful login and login failure audit events differ slightly depending on the underlying identity provider. There are the following cases to consider:
+
+1. **Password-based identity providers**
+
+All password-based identity providers are using a [central location](https://github.com/openshift/oauth-server/blob/690499e76a0b242adb5ccc73f23cccc8a50b8788/pkg/server/login/login.go#L173) where login events are handled. Both login success and failure events will be emitted here.
+
+2. **External OAuth identity providers**
+
+For login events external OAuth providers must invoke oauth-server callback handler to finalize authorization using [code grant flow](https://datatracker.ietf.org/doc/html/rfc6749#section-4.1).
+
+Login success will be emitted upon successful code exchange with oauth-server in the [central OAuth callback handler](https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L149-L157). 
+
+Login failure however cannot be easily detected as there is no guarantee that external oauth providers will invoke callback in the case of a login failure. Hence, only best-effort login failure events can be emitted, if the external OAuth provider provides an `error` response in the [central `HandleRequest` handler](https://github.com/openshift/oauth-server/blob/513a8bb5b6cbd5e8faacbed523eb861aa29a674b/vendor/github.com/RangelReale/osincli/authorize.go#L84).
+
+3. **External OAuth identity providers used as challengers**
```

Comment:
I didn't get that one here, but it might just me. The only thing I could find with regards to "challenge" in the given context was referring to PKCE.

Reply:
"challenger" is a specific terminology to the oauth-server, it is an IdP posting HTTP challenges (in this specific case the Basic-Auth HTTP Challenge). This section is basically the same as `1. Password-based identity providers`

Reply:
So what in particular is a challenge? The state that is returned to the callback endpoint? Do you have a link to look it up?

Reply:
This is something we add on oauth-server side right? If so, is it similar to PKCE to ensure the IDP has a token we have generated.

Reply:
@ibihim the code location is documented in this OEP, see the first paragraph (https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/oauth/external/handler.go#L118). If you look at the caller it simply takes credentials from basic auth as @stlaz explained: https://github.com/openshift/oauth-server/blob/822478f88514a80f053d9f65cd689d981b6ea4fd/pkg/authenticator/request/basicauthrequest/basicauth.go#L27

It effectively is https://oauth.net/2/grant-types/password/

Reply:
@EmilyM1 no, PKCE extends code grant flow to make it usable in public clients.